llarp/path/pathbuilder.cpp

```diff
@@ -292,44 +295,63 @@ namespace llarp
     }
 
     std::optional<std::vector<RouterContact>>
-    Builder::GetHopsAlignedToForBuild(RouterID endpoint)
+    Builder::GetHopsAlignedToForBuild(RouterID endpoint, std::set<RouterID> exclude)
     {
+      const auto pathConfig = m_router->GetConfig()->paths;
+
       std::vector<RouterContact> hops;
       {
-        const auto maybe = SelectFirstHop();
+        const auto maybe = SelectFirstHop(exclude);
         if (not maybe.has_value())
           return std::nullopt;
         hops.emplace_back(*maybe);
       };
+
+      RouterContact endpointRC;
+      if (const auto maybe = m_router->nodedb()->Get(endpoint))
+      {
+        endpointRC = *maybe;
+      }
+      else
+        return std::nullopt;
+
       for (size_t idx = hops.size(); idx < numHops; ++idx)
       {
         if (idx + 1 == numHops)
         {
-          const auto maybe = m_router->nodedb()->Get(endpoint);
-          if (maybe.has_value())
-          {
-            hops.emplace_back(*maybe);
-          }
-          else
-            return std::nullopt;
+          hops.emplace_back(endpointRC);
         }
         else
         {
-          const auto maybe = m_router->nodedb()->GetRandom(
-              [&hops, r = m_router, endpoint](const auto& rc) -> bool {
-                if (r->routerProfiling().IsBadForPath(rc.pubkey))
-                  return false;
-                for (const auto& hop : hops)
-                {
-                  if (hop.pubkey == rc.pubkey)
-                    return false;
-                }
-                return rc.pubkey != endpoint;
-              });
-
-          if (not maybe.has_value())
+          auto filter =
+              [&hops, r = m_router, endpointRC, pathConfig, exclude](const auto& rc) -> bool {
+            if (exclude.count(rc.pubkey))
+              return false;
+
+            std::set<RouterContact> hopsSet;
+            hopsSet.emplace(endpointRC);
+            for (const auto& hop : hops)
+              hopsSet.emplace(hop);
+
+            if (r->routerProfiling().IsBadForPath(rc.pubkey))
+              return false;
+            for (const auto& hop : hopsSet)
+            {
+              if (hop.pubkey == rc.pubkey)
+                return false;
+            }
+
+            hopsSet.emplace(rc);
+            if (not pathConfig.Acceptable(hopsSet))
+              return false;
+
+            return rc.pubkey != endpointRC.pubkey;
+          };
+
+          if (const auto maybe = m_router->nodedb()->GetRandom(filter))
```

Comment:
This has some non-uniform selection problems; discussion off github.

Reply:
Problem: GetRandom()'s implementation isn't random enough: if the filter excludes a big chunk of nodes then we have a good chance of selecting into that chunk, in which case we will heavily favour selection of the node that immediately follows the bad chunk.

Potential solution: make GetRandom shuffle a set of indices (or std::ref's, if we don't have random access into the list) and walk through those rather than walking sequentially.